scripts/foundry/contract-upgrades/v3.1.0/README.md

```diff
@@ -0,0 +1,118 @@
+# Nitro contracts 3.1.0 upgrade
+
+To upgrade to Nitro contracts 3.1.0 (BoLD), you will need to use the [BOLDUpgradeAction](https://github.com/EspressoSystems/nitro-contracts/blob/main/src/rollup/BOLDUpgradeAction.sol) from the [nitro-contract](https://github.com/EspressoSystems/nitro-contracts) repo.
+
+BOLDUpgradeAction will perform the following actions:
+
+1. Upgrade bridge, inbox, rollupEventInbox, outbox, sequencerInbox to v3.1.0
+2. Deploy new v3.1.0 BoLD challenge manager
+3. Migrate v2 rollup into a new v3.1.0 rollup address
+4. Setup the rollup according to the new config and use the latest confirmed assertion on the old rollup as genesis of the new rollup
+
+Note that contracts without code changes are not upgraded. It is normal to have some contracts still in the old version after the upgrade as they are equivalent to the new version. There are no associated ArbOS upgrade for this version.
+
+## Requirements
+
+**Nitro contracts**: This upgrade only support upgrading from the v2.1.3 version of Espresso nitro-contracts.
+
+## How to use it
+
+1. Clone nitro-contract repo
+
+```
+$ git clone https://github.com/EspressoSystems/nitro-contracts.git
+$ cd nitro-contracts
+```
+
+2. Checkout the v3.1.0-c6f2078 tag
+
+```
+$ git checkout v3.1.0-c6f2078
+```
+
+3. Install dependencies and build contracts
+
+```
+$ npx hardhat clean && forge clean
+$ yarn install
+$ yarn build:all
+```
+
+4. Edit `scripts/files/configs/custom.ts`, be very careful as they will override the existing configuration of your chain. Note: You will also have to add the `EspressoTEEVerifier` address. You can get it by calling the following command (replace `SequencerInboxProxyAddress` with the address of your SequencerInbox proxy):
+
+```
+$ cast call $SequencerInboxProxyAddress "espressoTEEVerifier()" --rpc-url $PARENT_CHAIN_RPC_URL
+```
+
+The `isDelayBufferable` should be set to `false` for EspressoIntegration.
+
+5. Setup .env in project root, make sure `CONFIG_NETWORK_NAME=custom`.
+
+```
+$ cp .env-sample .env
+```
+
+6. Run the prepare script, this will deploy the action with the specified configuration parameters. Note that `L1_PRIV_KEY` does not need to be the chain owner. Pass in the parent chain for `--network`; you can find the identifiers of these networks in the `hardhat.config.ts` file. Note that if your parent chain is an L1, you'll have to configure an additional `INFURA_KEY` env variable for its endpoint.
+   - Optionally, the script can try to verify the deployed contract by setting `DISABLE_VERIFICATION` to `false`. In that case, use the correct key for verifying the contracts on the block explorer depending on your parent chain: `ETHERSCAN_API_KEY | ARBISCAN_API_KEY | NOVA_ARBISCAN_API_KEY | BASESCAN_API_KEY`
```

Comment:
The Etherscan API is switching from V1 to [V2](https://docs.etherscan.io/etherscan-v2/v2-quickstart). You can take a look at [this](https://github.com/EspressoSystems/espresso-tee-contracts/pull/17) PR.

Reply:
Hmm, this is from offchain labs I didnt change it. I think for now, their code probably still requires API Key for all chains